### ISSUE TITLE
`dipy==1.8.0` causes `trx-python` dependency conflict with `ivadomed`, resulting in failing SCT installations

### ISSUE THREAD

**Issue by po09i** [spinalcordtoolbox/spinalcordtoolbox#4319]
### Description
I tried installing SCT and it is failing at the step when it's installing nilearn-0.6.2. I noticed the tests are also failing in the CI for all platforms in the repo.

### Steps to Reproduce
The [tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/workflows/tests.yml?query=branch%3Amaster) seem to have been failing for the pas 4 days. See [this test](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/7260966482/job/19781204087#step:4:538) for an example.

**Comment by jcohenadad** [comment]
Thank you for reporting this issue @po09i. I can reproduce it on my local station. Possible cause (and solution):

~~~
  Preparing metadata (setup.py): finished with status 'error'
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      The 'sklearn' PyPI package is deprecated, use 'scikit-learn'
      rather than 'sklearn' for pip commands.
      
      Here is how to fix this error in the main use cases:
      - use 'pip install scikit-learn' rather than 'pip install sklearn'
      - replace 'sklearn' by 'scikit-learn' in your pip requirements files
        (requirements.txt, setup.py, setup.cfg, Pipfile, etc ...)
      - if the 'sklearn' package is used by one of your dependencies,
        it would be great if you take some time to track which package uses
        'sklearn' instead of 'scikit-learn' and report it to their issue tracker
      - as a last resort, set the environment variable
        SKLEARN_ALLOW_DEPRECATED_SKLEARN_PACKAGE_INSTALL=True to avoid this error
      
      More information is available at
      https://github.com/scikit-learn/sklearn-pypi-package
      [end of output]
~~~

**Comment by jcohenadad** [comment]
Hum, nevermind. We already use `scikit-learn`:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/c7a8072fd63a06a2775a74029c042833f0fce510/requirements.txt#L43

Maybe the issue comes from another library (eg ivadomed?)

**Comment by jcohenadad** [comment]
Strangely, nilearn's install seem to point to "scikit-learn":
https://github.com/nilearn/nilearn/blob/8b0657bfe5338c63923323f3b2ff5c45b5a953d9/pyproject.toml#L32

So how come I'm seeing this during SCT's install:
~~~
Collecting sklearn (from nilearn->-r requirements.txt (line 20))
  Downloading sklearn-0.0.post12.tar.gz (2.6 kB)
~~~

**Comment by jcohenadad** [comment]
I tried running:
~~~
SKLEARN_ALLOW_DEPRECATED_SKLEARN_PACKAGE_INSTALL=True
~~~
Before installation, but that didn't make any difference.

**Comment by jcohenadad** [comment]
Workaround until this issue gets resolved: remove "nilearn" from the requirements.txt. This is a dirty workaround as related functionalities won't work, but in my case i'm not using nilearn and at least i'm able to install SCT successfully and continue my development on #4298

**Comment by po09i** [comment]
I think the issue with Nilearn is the result of Pip running into dependency conflicts and trying to resolve it by trying multiple inferior versions of the packages. See:

Collecting nilearn (from -r requirements.txt (line 20))
  Downloading nilearn-0.10.1-py3-none-any.whl (10.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.3/10.3 MB 14.4 MB/s eta 0:00:00
  Downloading nilearn-0.10.0-py3-none-any.whl (9.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 13.6 MB/s eta 0:00:00
  Downloading nilearn-0.9.2-py3-none-any.whl (9.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 13.6 MB/s eta 0:00:00
  Downloading nilearn-0.9.1-py3-none-any.whl (9.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 13.6 MB/s eta 0:00:00
  Downloading nilearn-0.9.0-py3-none-any.whl (10.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 13.7 MB/s eta 0:00:00
  Downloading nilearn-0.8.1-py3-none-any.whl (10.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 13.5 MB/s eta 0:00:00
  Downloading nilearn-0.8.0-py3-none-any.whl (4.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 13.2 MB/s eta 0:00:00
  Downloading nilearn-0.7.1-py3-none-any.whl (3.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 12.9 MB/s eta 0:00:00
  Downloading nilearn-0.7.0-py3-none-any.whl (3.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 13.1 MB/s eta 0:00:00
  Downloading nilearn-0.6.2-py3-none-any.whl (2.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 13.0 MB/s eta 0:00:00
Collecting sklearn (from nilearn->-r requirements.txt (line 20))
  Downloading sklearn-0.0.post12.tar.gz (2.6 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
  error: subprocess-exited-with-error
  
 Unclear to me what the conflict is though.

**Comment by po09i** [comment]
The problem seems to occur due to a new version of dipy (1.8.0). This new version requires trx-python>=0.2.9, however trx-python requires nibabel>5 which is incompatible with ivadomed (nibabel~=3.2).

I'm not familiar with the dependencies of SCT but **for now** limiting dipy to 1.5.0 (1.6.0 and 1.7.0 also caused issues according to the requirement.txt file) solved the dependency problem.

**Comment by jcohenadad** [comment]
Thank you @po09i , this is a much better temp fix than my suggestion to get rid of nilearn 😅

**Comment by po09i** [comment]
Any chance we can do a quick PR for this? Tests are failing in ST because SCT is required to test all the dependencies.

**Comment by jcohenadad** [comment]
both Joshua and Mathieu are on holidays. I see more pros than cons in doing the fix you propose (pin dipy to 1.5.x) so that your project can move forward (and when they come back, they will assess what the best strategy is). Can you implement the fix in a PR and I'll approve it?

**Comment by joshuacwnewton** [comment]
> The problem seems to occur due to a new version of dipy (1.8.0). This new version requires trx-python>=0.2.9, however trx-python requires nibabel>5 which is incompatible with ivadomed (nibabel~=3.2) and possibly other packages but I stopped digging here.

Thank you for this investigation! We've run into similar issues in the past related to the ivadomed project aggressively restricting its own dependencies, thus causing conflicts downstream for SCT and/or ADS.

I think it would be worthwhile to go through all of ivadomed's dependencies that are currently marked with `~=` and see if we can upgrade them where possible. Then, consider a new strategy for maintaining ivadomed's dependencies that would require the least amount of manual intervention long-term, given that `ivadomed` is largely unmaintained (apart for some minimal work to keep the package functional).

(For example, we could pin only the major version of packages in ivadomed's `requirements.txt`? Or, alternatively, maybe we could leave dependencies unpinned on the ivadomed side of things, and only pin them on the SCT/ADS side of things, to avoid needing to repeatedly create new `ivadomed` releases? Not sure...)
